How many late days are you using for this assignment?
0 

Name: <redacted> <redacted>
    Commit ID: 1a2ef6af9422d33afef4da24f19b62c9d7d66bd7

In [131]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.preprocessing import MinMaxScaler

# Part 1: Load the dataset

In [132]:
# Load the given datasets
data_catagorical = pd.read_csv(r'C:\Users\deshn\Documents\CS383\assignment-3-data-cleaning-and-preprocessing-<redacted>\chronic_kidney_disease_categorical.csv')
data_numerical = pd.read_csv(r'C:\Users\deshn\Documents\CS383\assignment-3-data-cleaning-and-preprocessing-<redacted>\chronic_kidney_disease_numerical.csv')
# Print the data
print(data_catagorical)
print(data_numerical)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:
Both the datasets help with the prediction of chronic kidney disease. One of the datasets has columns with numeric values and the other has columns with categorical values (like yes or no categories). Each unique ID corresponds to a patient, information about the patient is stored in the columns. The label of the dataset is the target column which is class, as it specifies if a patient has chronic kidney disease or not. All the other columns in both the datasets are features, which are the attributes of the patient. Like age, bp, rbc, rbcc etc.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [133]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
num=data_numerical.duplicated(subset= ['unique_id']).sum()
# Drop these duplicate rows
no_duplicates_num=data_numerical.drop_duplicates(subset= ['unique_id'])
print(num)

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
data_catagorical.duplicated(subset= ['unique_id'])
duplicates=data_catagorical[data_catagorical.duplicated(subset= ['unique_id'])]
print(duplicates)
no_duplicates_cat=data_catagorical.drop_duplicates(subset= ['unique_id'])


30
     unique_id   al   su       rbc        pc         pcc          ba  htn  \
231     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   
401     628271  2.0  1.0  abnormal    normal  notpresent  notpresent  yes   
402     158929  1.0  0.0       NaN       NaN  notpresent  notpresent   no   
403     240975  0.0  0.0    normal    normal  notpresent  notpresent   no   
404     441170  0.0  0.0    normal    normal  notpresent  notpresent   no   
405     337834  4.0  0.0    normal  abnormal     present  notpresent  yes   
406     118191  0.0  0.0    normal    normal  notpresent  notpresent   no   
407     224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
408     302906  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
409     938027  0.0  0.0    normal    normal  notpresent  notpresent   no   
410     572140  0.0  0.0    normal    normal  notpresent  notpresent   no   
411     283427  0.0  0.0    normal    normal  notpresent  notpresent   no

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:
No, not all the unique IDs are present in both datasets. Because the two datasets may have information about some people not present in the other dataset. So the rows that are missing in one dataset will show all the values for those columns to be NA and still be in the combined dataset.

In [134]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
merged_data = pd.merge(no_duplicates_cat, no_duplicates_num, how='outer', on='unique_id')

#Print the combined dataset
print(merged_data)

     unique_id   al   su     rbc        pc         pcc          ba  htn   dm  \
0       102677  3.0  0.0  normal  abnormal  notpresent  notpresent  yes   no   
1       103570  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
2       104675  2.0  1.0  normal    normal     present  notpresent   no  yes   
3       105923  1.0  0.0     NaN    normal  notpresent  notpresent  yes  yes   
4       107297  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
..         ...  ...  ...     ...       ...         ...         ...  ...  ...   
395     995177  4.0  3.0  normal  abnormal     present     present  yes  yes   
396     996076  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
397     996412  NaN  NaN     NaN       NaN  notpresent  notpresent   no   no   
398     997629  0.0  0.0  normal    normal         NaN         NaN   no   no   
399     997876  0.0  0.0     NaN    normal  notpresent  notpresent   no   no   

     cad appet   pe  ane  Target   age 

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [135]:
# Calculate the percentage of rows that contain atleast one missing value
null_value_row = merged_data.isna().any(axis=1).sum()
sum = len(merged_data)
percentage = (null_value_row)*100/sum
# Print %
print(percentage)
# Drop these rows from the dataset
merged_full_data = merged_data.dropna()
# Print the Dataset
print(merged_full_data)

61.75
     unique_id   al   su     rbc        pc         pcc          ba  htn   dm  \
0       102677  3.0  0.0  normal  abnormal  notpresent  notpresent  yes   no   
6       109053  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
7       114717  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
8       118191  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
9       118596  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
..         ...  ...  ...     ...       ...         ...         ...  ...  ...   
390     980291  3.0  0.0  normal  abnormal  notpresent  notpresent  yes  yes   
391     991031  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
394     994377  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
395     995177  4.0  3.0  normal  abnormal     present     present  yes  yes   
396     996076  0.0  0.0  normal    normal  notpresent  notpresent   no   no   

     cad appet   pe  ane  Target 

## Part 3.4: Sort the dataset according to the Labels

In [136]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_dataset = merged_full_data.sort_values(by=['Target']).reset_index(drop=True)

# Print the dataset
print(sorted_dataset)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       102677  3.0  0.0    normal  abnormal  notpresent  notpresent  yes   
1       337834  4.0  0.0    normal  abnormal     present  notpresent  yes   
2       343710  2.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       349892  1.0  3.0  abnormal  abnormal  notpresent  notpresent  yes   
4       353025  3.0  4.0    normal    normal  notpresent  notpresent  yes   
..         ...  ...  ...       ...       ...         ...         ...  ...   
148     486397  0.0  0.0    normal    normal  notpresent  notpresent   no   
149     203466  0.0  0.0    normal    normal  notpresent  notpresent   no   
150     482105  0.0  0.0    normal    normal  notpresent  notpresent   no   
151     228849  0.0  0.0    normal    normal  notpresent  notpresent   no   
152     996076  0.0  0.0    normal    normal  notpresent  notpresent   no   

      dm  cad appet   pe  ane  Target   age    bp    bgr     bu    sc    so

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [137]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
encoded_dataset = pd.get_dummies(sorted_dataset,columns=categorical_columns, dtype= int)

# Print the dataset
print(encoded_dataset)

     unique_id   al   su   age    bp    bgr     bu    sc    sod  pot  hemo  \
0       102677  3.0  0.0  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1   
1       337834  4.0  0.0  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2   
2       343710  2.0  0.0  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7   
3       349892  1.0  3.0  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6   
4       353025  3.0  4.0  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9   
..         ...  ...  ...   ...   ...    ...    ...   ...    ...  ...   ...   
148     486397  0.0  0.0  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6   
149     203466  0.0  0.0  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5   
150     482105  0.0  0.0  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0   
151     228849  0.0  0.0  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0   
152     996076  0.0  0.0  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7   

      pcv     wbcc  rbcc  rbc_abnormal  rbc_normal  pc_abnormal

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: 
It will not be the same in the downstream models, it uses the one-hot encoding to convert the categorical variable into a numerical variable and then is able to use them as independent variables in the model and learn how each of them is related to the target variable.
If the catagorical variable can take on more than 2 values, then it will be one-hot encoded. It will create 5 seperate columns for each value and put a 1 in the column that is in the sandwitch and 0 otherwise.


## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [138]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
for column in numerical_columns:
    mean = encoded_dataset[column].mean()
    std = encoded_dataset[column].std()
    lower = mean - 3*std
    upper = mean + 3*std

    encoded_dataset = encoded_dataset[(encoded_dataset[column] >= lower) & (encoded_dataset[column] <= upper)]
    
# Print the dataset
print(encoded_dataset)

     unique_id   al   su   age     bp    bgr     bu   sc    sod  pot  hemo  \
2       343710  2.0  0.0  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7   
5       397388  2.0  2.0  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0   
7       438182  2.0  0.0  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5   
8       474407  4.0  0.0  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3   
9       475200  4.0  2.0  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9   
..         ...  ...  ...   ...    ...    ...    ...  ...    ...  ...   ...   
148     486397  0.0  0.0  20.0   70.0  123.0   44.0  1.0  135.0  3.8  14.6   
149     203466  0.0  0.0  22.0   60.0   97.0   18.0  1.2  138.0  4.3  13.5   
150     482105  0.0  0.0  44.0   60.0   95.0   46.0  0.5  138.0  4.2  15.0   
151     228849  0.0  0.0  58.0   80.0  100.0   50.0  1.2  140.0  3.5  14.0   
152     996076  0.0  0.0  57.0   60.0  105.0   49.0  1.2  150.0  4.7  15.7   

      pcv     wbcc  rbcc  rbc_abnormal  rbc_normal  pc_abnormal

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [139]:
# Normalize the all Numerical Attributes in the dataset.
scaler = MinMaxScaler()
normalised = encoded_dataset.copy()
normalised[numerical_columns] = scaler.fit_transform(normalised[numerical_columns])
#data_minmax = scaler.fit_transform(encoded_dataset[numerical_columns])

# Print the dataset
#normalised= pd.DataFrame(data_minmax, columns=numerical_columns)
print(normalised)
#print(pd.DataFrame(data_minmax, columns=numerical_columns))


     unique_id   al   su       age    bp       bgr        bu        sc  \
2       343710  2.0  0.0  0.743243  0.50  0.442060  0.901961  0.432099   
5       397388  2.0  2.0  0.770270  1.00  0.901288  0.163399  0.345679   
7       438182  2.0  0.0  0.905405  0.50  0.785408  0.862745  0.518519   
8       474407  4.0  0.0  0.202703  0.75  0.158798  0.196078  0.160494   
9       475200  4.0  2.0  0.783784  1.00  0.399142  0.287582  0.839506   
..         ...  ...  ...       ...   ...       ...       ...       ...   
148     486397  0.0  0.0  0.189189  0.25  0.227468  0.222222  0.074074   
149     203466  0.0  0.0  0.216216  0.00  0.115880  0.052288  0.098765   
150     482105  0.0  0.0  0.513514  0.00  0.107296  0.235294  0.012346   
151     228849  0.0  0.0  0.702703  0.50  0.128755  0.261438  0.098765   
152     996076  0.0  0.0  0.689189  0.00  0.150215  0.254902  0.098765   

          sod       pot      hemo       pcv      wbcc      rbcc  rbc_abnormal  \
2    0.500000  0.793103  0.000

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them.




Answer:
There are 12 columns in the dataset which are not appropriate for modeling and predictions. The columns which are not appropriate for modeling and predictions are:

'unique_id'- As it can lead to overfitting of the model and it does not give any important information for prediction.

All these columns can be removed from the dataset.
As they are all reduntand because they give the same information as the counter columns.eq- rbc_abnormal and so on.
------'rbc_normal',
'pc_normal',
'pcc_notpresent'
'ba_notpresent',
'htn_no',
'dm_no',
'cad_no',
'appet_good',
'pe_no',
'ane_no',
'Target_notckd'

In [140]:
#Remove that column
removed_dataset = normalised.drop(columns=['unique_id','rbc_normal','pc_normal','pcc_notpresent','ba_notpresent','htn_no','dm_no','cad_no','appet_good','pe_no','ane_no','Target_notckd'])

# Print the dataset
print(removed_dataset.head(15))

     al   su       age    bp       bgr        bu        sc       sod  \
2   2.0  0.0  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
5   2.0  2.0  0.770270  1.00  0.901288  0.163399  0.345679  0.766667   
7   2.0  0.0  0.905405  0.50  0.785408  0.862745  0.518519  0.600000   
8   4.0  0.0  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
9   4.0  2.0  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
10  3.0  1.0  0.729730  0.00  0.935622  0.169935  0.160494  0.333333   
12  2.0  0.0  0.675676  0.75  0.253219  0.633987  0.777778  0.366667   
13  4.0  3.0  0.851351  0.25  0.618026  0.562092  0.728395  0.000000   
15  1.0  0.0  0.540541  0.00  0.399142  0.535948  0.358025  0.700000   
18  3.0  0.0  0.756757  0.25  0.223176  0.209150  0.160494  0.533333   
19  3.0  1.0  0.662162  0.50  0.618026  0.411765  0.432099  0.566667   
20  4.0  1.0  0.783784  0.00  0.725322  0.313725  0.481481  0.566667   
22  4.0  0.0  0.878378  0.00  0.206009  0.751634  0.604938  0.53

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [141]:
# Export the dataframe to a new csv file
removed_dataset.to_csv('cleaned_dataset.csv')
# Export the dataframe to a new json file
removed_dataset.to_json('cleaned_dataset.json')

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

Paste here: 

In [142]:
clean_data_csv = pd.read_csv(r'C:\Users\deshn\Documents\CS383\assignment-3-data-cleaning-and-preprocessing-<redacted>\cleaned_dataset.csv')
pd.set_option('display.max_columns', None)

clean_data_csv.head(15)

,Unnamed: 0,al,su,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,rbc_abnormal,pc_abnormal,pcc_present,ba_present,htn_yes,dm_yes,cad_yes,appet_poor,pe_yes,ane_yes,Target_ckd
0,2,2.0,0.0,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.793103,0.000000,0.032258,0.395161,0.057143,1,1,0,0,1,1,1,1,1,1,1
1,5,2.0,2.0,0.770270,1.00,0.901288,0.163399,0.345679,0.766667,0.206897,0.524752,0.548387,0.443548,0.342857,0,0,0,1,1,0,1,0,0,0,1
2,7,2.0,0.0,0.905405,0.50,0.785408,0.862745,0.518519,0.600000,1.000000,0.277228,0.322581,0.233871,0.371429,1,1,0,0,1,1,1,0,0,0,1
3,8,4.0,0.0,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.206897,0.059406,0.000000,0.653226,0.257143,0,1,1,1,0,0,0,0,0,1,1
4,9,4.0,2.0,0.783784,1.00,0.399142,0.287582,0.839506,0.666667,0.586207,0.019802,0.096774,0.258065,0.114286,1,1,0,1,1,1,0,0,1,0,1
5,10,3.0,1.0,0.729730,0.00,0.935622,0.169935,0.160494,0.333333,0.034483,0.019802,0.064516,0.879032,0.000000,0,1,1,0,1,0,0,1,0,1,1
6,12,2.0,0.0,0.675676,0.75,0.253219,0.633987,0.777778,0.366667,0.655172,0.138614,0.193548,0.169355,0.114286,1,1,0,0,1,0,0,0,0,0,1
7,13,4.0,3.0,0.851351,0.25,0.618026,0.562092,0.728395,0.000000,0.344828,0.168317,0.161290,0.580645,0.085714,0,1,1,1,1,1,1,0,1,1,1
8,15,1.0,0.0,0.540541,0.00,0.399142,0.535948,0.358025,0.700000,0.379310,0.207921,0.161290,0.830645,0.057143,0,0,0,0,1,1,0,0,0,0,1
9,18,3.0,0.0,0.756757,0.25,0.223176,0.209150,0.160494,0.533333,0.620690,0.485149,0.516129,0.290323,0.257143,0,1,0,0,1,1,0,0,0,0,1


| Unnamed: 0 | al  | su  |    age   |    bp    |    bgr    |    bu    |    sc    |   sod    |   pot    |   hemo   |   pcv    |  wbcc    |  rbcc    |  rbc_abnormal |  pc_abnormal |  pcc_present |  ba_present |  htn_yes |  dm_yes |  cad_yes |  appet_poor |  pe_yes |  ane_yes | Target_ckd |
|------------|-----|-----|----------|----------|-----------|----------|----------|----------|----------|----------|----------|----------|----------|----------------|---------------|--------------|-------------|----------|---------|----------|--------------|---------|----------|-------------|
| 0          | 2.0 | 0.0 | 0.743243 | 0.50     | 0.442060  | 0.901961 | 0.432099 | 0.500000 | 0.793103 | 0.000000 | 0.032258 | 0.395161 | 0.057143 | 1              | 1             | 0            | 0           | 1        | 1       | 1        | 1            | 1       | 1        | 1           |
| 1          | 2.0 | 2.0 | 0.770270 | 1.00     | 0.901288  | 0.163399 | 0.345679 | 0.766667 | 0.206897 | 0.524752 | 0.548387 | 0.443548 | 0.342857 | 0              | 0             | 0            | 1           | 1        | 0       | 1        | 0            | 0       | 0        | 1           |
| 2          | 2.0 | 0.0 | 0.905405 | 0.50     | 0.785408  | 0.862745 | 0.518519 | 0.600000 | 1.000000 | 0.277228 | 0.322581 | 0.233871 | 0.371429 | 1              | 1             | 0            | 0           | 1        | 1       | 1        | 0            | 0       | 0        | 1           |
| 3          | 4.0 | 0.0 | 0.202703 | 0.75     | 0.158798  | 0.196078 | 0.160494 | 0.166667 | 0.206897 | 0.059406 | 0.000000 | 0.653226 | 0.257143 | 0              | 1             | 1            | 1           | 0        | 0       | 0        | 0            | 0       | 1        | 1           |
| 4          | 4.0 | 2.0 | 0.783784 | 1.00     | 0.399142  | 0.287582 | 0.839506 | 0.666667 | 0.586207 | 0.019802 | 0.096774 | 0.258065 | 0.114286 | 1              | 1             | 0            | 1           | 1        | 1       | 0        | 0            | 1       | 0        | 1           |
| 5          | 3.0 | 1.0 | 0.729730 | 0.00     | 0.935622  | 0.169935 | 0.160494 | 0.333333 | 0.034483 | 0.019802 | 0.064516 | 0.879032 | 0.000000 | 0              | 1             | 1            | 0           | 1        | 0       | 0        | 1            | 0       | 1        | 1           |
| 6          | 2.0 | 0.0 | 0.675676 | 0.75     | 0.253219  | 0.633987 | 0.777778 | 0.366667 | 0.655172 | 0.138614 | 0.193548 | 0.169355 | 0.114286 | 1              | 1             | 0            | 0           | 1        | 0       | 0        | 0            | 0       | 0        | 1           |
| 7          | 4.0 | 3.0 | 0.851351 | 0.25     | 0.618026  | 0.562092 | 0.728395 | 0.000000 | 0.344828 | 0.168317 | 0.161290 | 0.580645 | 0.085714 | 0              | 1             | 1            | 1           | 1        | 1       | 1        | 0            | 1       | 1        | 1           |
| 8          | 1.0 | 0.0 | 0.540541 | 0.00     | 0.399142  | 0.535948 | 0.358025 | 0.700000 | 0.379310 | 0.207921 | 0.161290 | 0.830645 | 0.057143 | 0              | 0             | 0            | 0           | 1        | 1       | 0        | 0            | 0       | 0        | 1           |
| 9          | 3.0 | 0.0 | 0.756757 | 0.25     | 0.223176  | 0.209150 | 0.160494 | 0.533333 | 0.620690 | 0.485149 | 0.516129 | 0.290323 | 0.257143 | 0              | 1             | 0            | 0           | 1        | 1       | 0        | 0            | 0       | 0        | 1           |
| 10         | 3.0 | 1.0 | 0.662162 | 0.50     | 0.618026  | 0.411765 | 0.432099 | 0.566667 | 0.689655 | 0.316832 | 0.354839 | 0.250000 | 0.200000 | 0              | 1             | 1            | 1           | 1        | 1       | 0        | 0            | 1       | 0        | 1           |
| 11         | 4.0 | 1.0 | 0.783784 | 0.00     | 0.725322  | 0.313725 | 0.481481 | 0.566667 | 0.862069 | 0.178218 | 0.193548 | 0.258065 | 0.114286 | 1              | 1             | 0            | 1           | 1        | 1       | 0        | 1            | 1       | 0        | 1           |
| 12         | 4.0 | 0.0 | 0.878378 | 0.00     | 0.206009  | 0.751634 | 0.604938 | 0.533333 | 0.689655 | 0.366337 | 0.387097 | 0.879032 | 0.371429 | 0              | 0             | 0            | 0           | 1        | 1       | 0        | 1            | 1       | 0        | 1           |
| 13         | 3.0 | 0.0 | 0.878378 | 0.25     | 0.639485  | 0.470588 | 0.395062 | 0.433333 | 0.517241 | 0.267327 | 0.322581 | 0.104839 | 0.171429 | 0              | 1             | 1            | 1           | 1        | 1       | 1        | 0            | 0       | 0        | 1           |
| 14         | 1.0 | 0.0 | 0.716216 | 0.50     | 1.000000  | 0.163399 | 0.111111 | 0.066667 | 0.206897 | 0.267327 | 0.387097 | 0.532258 | 0.371429 | 1              | 0             | 0            | 0           | 0        | 1       | 0        | 1            | 0       | 0        | 1           |


### Paste the markdown table here

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [143]:
# Code to rename all the columns in the dataset
rename_dict = {
    'al': 'albumin',
    'su': 'sugar',
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'rbc_abnormal': 'rbc abnormal',
    'pc_abnormal': 'pus cell abnormal',
    'pcc_present': 'pus cell clumps present',
    'ba_present': 'bacteria present',
    'htn_yes': 'hypertension',
    'dm_yes': 'diabetes mellitus',
    'cad_yes': 'coronary artery disease',
    'appet_poor': 'appetite poor',
    'pe_yes': 'pedal edema',
    'ane_yes': 'anemia',
    'Target_ckd': 'class'
}

# Rename the columns
removed_dataset.rename(columns=rename_dict, inplace=True)

# Display the updated DataFrame
print(removed_dataset.head())

   albumin  sugar       age  blood pressure  blood glucose random  blood urea  \
2      2.0    0.0  0.743243            0.50              0.442060    0.901961   
5      2.0    2.0  0.770270            1.00              0.901288    0.163399   
7      2.0    0.0  0.905405            0.50              0.785408    0.862745   
8      4.0    0.0  0.202703            0.75              0.158798    0.196078   
9      4.0    2.0  0.783784            1.00              0.399142    0.287582   

   serum creatinine    sodium  potassium  hemoglobin  packed cell volume  \
2          0.432099  0.500000   0.793103    0.000000            0.032258   
5          0.345679  0.766667   0.206897    0.524752            0.548387   
7          0.518519  0.600000   1.000000    0.277228            0.322581   
8          0.160494  0.166667   0.206897    0.059406            0.000000   
9          0.839506  0.666667   0.586207    0.019802            0.096774   

       wbcc      rbcc  rbc abnormal  pus cell abnormal  

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [144]:
# Converted SQL to Pandas code
result = removed_dataset.groupby('class').size().reset_index(name='count')

# Display the result
print(result)

   class  count
0      0    111
1      1     22


The output of my code is providing the total number of total positive and negative results regarding kidney disease as 1's and 0's respectivly, from the given data. It is counting the occurances of each catagory of result from the given data. It shows that there are 22 positive results and 111 negative results.